src/epm.rs

```diff
@@ -219,47 +271,62 @@ where
 		.await?
 		.map(|score| score.0);
 
-	let voters = snapshot.voters.clone();
-	let targets = snapshot.targets.clone();
-
-	log::trace!(
-		target: LOG_TARGET,
-		"mine solution: desired_targets={}, voters={}, targets={}",
-		desired_targets,
-		voters.len(),
-		targets.len()
-	);
-
-	let blocking_task = tokio::task::spawn_blocking(move || match solver {
-		Solver::SeqPhragmen { iterations } => {
-			BalanceIterations::set(iterations);
-			Miner::<T>::mine_solution_with_snapshot::<
-				SequentialPhragmen<AccountId, Accuracy, Balancing>,
-			>(voters, targets, desired_targets)
-		},
-		Solver::PhragMMS { iterations } => {
-			BalanceIterations::set(iterations);
-			Miner::<T>::mine_solution_with_snapshot::<PhragMMS<AccountId, Accuracy, Balancing>>(
-				voters,
-				targets,
-				desired_targets,
-			)
-		},
-	})
-	.await;
-
-	match blocking_task {
-		Ok(Ok((solution, score, solution_or_snapshot_size))) => Ok(MinedSolution {
-			round,
-			desired_targets,
-			snapshot,
-			minimum_untrusted_score,
-			solution,
-			score,
-			solution_or_snapshot_size,
-		}),
-		Ok(Err(err)) => Err(Error::Other(format!("{:?}", err))),
-		Err(err) => Err(err.into()),
+	let mut voters = TrimmedVoters::new(snapshot.voters.clone());
+
+	loop {
+		let s = solver.clone();
+		let v = voters.get();
+		let t = snapshot.targets.clone();
+
+		let blocking_task =
+			tokio::task::spawn_blocking(move || match s {
+				Solver::SeqPhragmen { iterations } => {
+					BalanceIterations::set(iterations);
+					Miner::<T>::mine_solution_with_snapshot::<
+						SequentialPhragmen<AccountId, Accuracy, Balancing>,
+					>(v, t, desired_targets)
+				},
+				Solver::PhragMMS { iterations } => {
+					BalanceIterations::set(iterations);
+					Miner::<T>::mine_solution_with_snapshot::<
+						PhragMMS<AccountId, Accuracy, Balancing>,
+					>(v, t, desired_targets)
+				},
+			})
+			.await;
+
+		match blocking_task {
+			Ok(Ok((solution, score, solution_or_snapshot_size, t))) if !t.is_trimmed() => {
+				if solution.unique_targets().len() != desired_targets as usize {
+					return Err(Error::Feasibility(format!(
+						"Invalid winner count {}, expected {desired_targets}",
+						solution.unique_targets().len()
+					)))
+				}
+
+				return Ok(MinedSolution {
+					round,
+					desired_targets,
+					snapshot,
+					minimum_untrusted_score,
+					solution,
+					score,
+					solution_or_snapshot_size,
+				})
+			},
+			Ok(Ok((solution, _score, _, _))) => {
+				if solution.unique_targets().len() != desired_targets as usize {
+					return Err(Error::Feasibility(format!(
+						"Invalid winner count {}, expected {desired_targets}",
+						solution.unique_targets().len()
+					)))
+				}
+
+				voters.trim_next()?;
```

Comment:
As per my comment above, perhaps we could pass the number of voters to trim which will depend on the iteration we are at. E.g. if this is the first call to `fn trim_next` in the loop, trim N voters. After a few attempts, start to trim only one voter since we're approaching the solution that does not need any more trimming.